File: gfg/NextGreaterElementII.java
import java.util.*;

class NextGreaterElementII {
    public int[] nextGreaterElements(int[] nums) {
        ArrayList<Integer> arr = new ArrayList<>();
        for (int ele : nums)
            arr.add(ele);
        Stack<Integer> stk = new Stack<>();
        int[] sol = new int[nums.length];
        Arrays.fill(sol, -1);
        arr.addAll(arr);

        for (int i = 0; i < arr.size(); i++) {
            while (!stk.empty() && arr.get(stk.peek()) < arr.get(i)) {
                sol[stk.peek() % nums.length] = arr.get(i);
                stk.pop();
            }
            stk.push(i);
        }
        return sol;
    }

    public static void main(String[] args) {
        NextGreaterElementII obj = new NextGreaterElementII();
        int[] arr = { 1, 2, 1 };
        int[] sol = obj.nextGreaterElements(arr);
        System.out.println(Arrays.toString(sol));
    }
}

File: gfg/nextGreaterElementII.cpp
// https://leetcode.com/problems/next-greater-element-ii/solution/
// NEXT GREATER ELEMENT (NGE)
// TIME: O(n)
// SPACE: O(n)
// COMMENTS: Use stack and solve the problem of next element

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  vector<int> nextGreaterElements(vector<int> &nums) {
    stack<int> stk;
    vector<int> sol(nums.size(), -1);
    int totlen = nums.size();
    for (int i = 0; i < totlen * 2; i++) {
      while (!stk.empty() && nums[i % totlen] > nums[stk.top()]) {
        sol[stk.top()] = nums[i % totlen];
        stk.pop();
      }
      stk.push(i % totlen);
    }
    return sol;
  }
};


int main(int argc, char** argv) {
  Solution *sol = new Solution();
  vector<int> arr = {1, 2, 1};
  vector<int> ans = sol->nextGreaterElements(arr);

  for (int el: ans) {
    cout << el << " ";
  }
  cout << endl;
}